### PR TITLE
Design description

### PR DESCRIPTION
**Pull request by ohler55**
The plan it to point people to this article as a way of describing how the package was developed.

### CODE REVIEW

File: design.md [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+length is a great way to avoid allocating additional memory. Care has
+to be taken when collapsing though as any cells in the slice that
+point to objects will now leave those objects dangling or rather
+referenced but not reachable and they will never be garbage
+collected. Simply setting the slice slot to `nil` will avoid memory
+leaks.
+
+### Memory Allocation
+
+Like most languages, memory allocation adds overhead. Best to avoid
```
Comment (zuern) [inline design.md:445]:
```suggestion
Like most languages, memory allocation adds overhead. It's best to avoid
```

**Comment by ohler55** [comment]
I was wondering if more code samples or pictures would help. If so where might they be helpful?

**Comment by zuern** [comment]
Many more suggestions in the PR, don't forget to "Show hidden conversations".

**Comment by zuern** [comment]
> I was wondering if more code samples or pictures would help. If so where might they be helpful?

Perhaps show an equivalent JSONPath being constructed by parsing a string?

**Pull request merged**